File: core/src/lib/WebAdapter.ts
export interface WebRequest {
  body: any;
  headers: Record<string, string>;
  method: string;
}

export interface WebResponse {
  write(data: string): void;
  end(data?: string): void;
  setHeader(name: string, value: string): void;
  status(code: number): WebResponse;
}

export interface WebStreamHandler {
  onChunk(chunk: string): Promise<void>;
  onComplete?(): Promise<void>;
  onError?(error: Error): Promise<void>;
}

export class WebOutputAdapter {
  private response: WebResponse;
  private streamHandler?: WebStreamHandler;

  constructor(response: WebResponse, streamHandler?: WebStreamHandler) {
    this.response = response;
    this.streamHandler = streamHandler;
  }

  async handleOutput(output: any): Promise<void> {
    if (this.streamHandler) {
      if (typeof output === 'string') {
        await this.streamHandler.onChunk(output);
      } else {
        await this.streamHandler.onChunk(JSON.stringify(output) + '\n');
      }
    } else {
      if (typeof output === 'string') {
        this.response.write(output);
      } else {
        this.response.write(JSON.stringify(output) + '\n');
      }
    }
  }

  async handleError(error: Error): Promise<void> {
    if (this.streamHandler?.onError) {
      await this.streamHandler.onError(error);
    } else {
      const errorResponse = JSON.stringify({ error: error.message });
      this.response.status(500);
      this.response.end(errorResponse);
    }
  }

  async complete(): Promise<void> {
    if (this.streamHandler?.onComplete) {
      await this.streamHandler.onComplete();
    } else {
      this.response.end();
    }
  }
}

File: core-python/src/article_dryer/lib/web_adapter.py
from typing import Any, Optional, Protocol, Union
from fastapi import Response
from pydantic import BaseModel

class WebRequest(BaseModel):
    body: Union[str, dict]
    headers: dict = {}

class WebResponse(Protocol):
    async def write(self, data: bytes) -> None: ...
    async def end(self) -> None: ...

class WebStreamHandler(Protocol):
    async def handle(self, data: Any) -> None: ...

class WebOutputAdapter:
    def __init__(self, response: WebResponse, stream_handler: Optional[WebStreamHandler] = None):
        self.response = response
        self.stream_handler = stream_handler

    async def handle_output(self, data: Any) -> None:
        if self.stream_handler:
            await self.stream_handler.handle(data)
        else:
            await self.response.write(str(data).encode())

    async def handle_error(self, error: Exception) -> None:
        error_data = {
            'error': str(error),
            'type': error.__class__.__name__
        }
        await self.response.write(str(error_data).encode())
        await self.complete()

    async def complete(self) -> None:
        await self.response.end()
